<div style="text-align: right"> Mark Hendricks </div>

<left>FINM 37500 - Fixed Income Derivatives</left> 
<br>
<left>Winter 2024</left>

<h2><center> Homework #3- Modeling Volatility, Pricing w/ BDT, Midcurve Swaptions </center></h2>

<center>Due: Wednesday, Feb 28 at 6PM</center>

<h3><span style="color:#00008B">Name - Nick Lewis</span></h3>
<h3><span style="color:#00008B">Email - nicklewis16@uchicago.edu</span></h3>

# Homework 3

## FINM 37500: Fixed Income Derivatives

### Mark Hendricks

#### Winter 2024

In [130]:
import numpy as np
import pandas as pd
import datetime
import warnings
from scipy.stats import norm
from scipy.optimize import fsolve

import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
plt.rcParams['figure.figsize'] = (12,6)
plt.rcParams['font.size'] = 15
plt.rcParams['legend.fontsize'] = 13

from matplotlib.ticker import (MultipleLocator,
                               FormatStrFormatter,
                               AutoMinorLocator)
import warnings
warnings.filterwarnings('ignore')


import sys
sys.path.insert(0, '../cmds')
from my_cmds import *

# 1. Modeling the Volatility Smile

## Swaption Vol Data

The file `data/swaption_vol_data_2024-02-20.xlsx` has market data on the implied volatility skews for swaptions. Note that it has several columns:
* `expry`: expiration of the swaption
* `tenor`: tenor of the underlying swap
* `model`: the model by which the volatility is quoted. (All are Black.)
* `-200`, `-100`, etc.: The strike listed as difference from ATM strike (bps). Note that ATM is considered to be the **forward swapa rate** which you can calculate.

In [131]:
swaption_vol = pd.read_excel(f'../data/swaption_vol_data.xlsx')

freqcurve = 4
RELATIVE_STRIKE = 0
N = 100

SWAP_TYPE = 'SOFR'
QUOTE_STYLE = 'black'
expry = 1
tenor = 4
DATE = '2024-02-20'

isPayer = True
freqswap = 4
swaption_vol

reference instrument  model       date  expiration  tenor   -200   -100  \
0      SOFR   swaption  black 2024-02-20           1      4  54.54  40.37   

     -50    -25      0     25     50    100    200  
0  35.94  34.23  32.83  31.71  30.86  29.83  29.54

Your data: ywill use a single row of this data for the `1x4` swaption.
* date: `2024-02-20`
* expiration: 1yr
* tenor: 4yrs

## Rate Data

The file `data/cap_quotes_2024-02-20.xlsx` gives 
* SOFR swap rates, 
* their associated discount factors
* their associated forward interest rates.

You will not need the cap data (flat or forward vols) for this problem.
* This cap data would be helpful in calibrating a binomial tree, but this problem focuses on Black's formula and SABR.

## The Swaption

Consider the following swaption with the following features:
* underlying is a fixed-for-floating (SOFR) swap
* the underlying swap has **quarterly** payment frequency
* this is a **payer** swaption, which gives the holder the option to **pay** the fixed swap rate and receive SOFR.

In [132]:
FILEDATE = '2024-02-20'
FILEIN = f'../data/cap_quotes_{FILEDATE}'
BB_COMPOUND = 1 #compounding of quoted SOFR swaps
freqcap = 4
DATE = '2024-02-20'

In [133]:
## QUOTES TO CURVES
SHEET = 'sofr'
sofrdata_raw = pd.read_excel(f'{FILEIN}.xlsx', sheet_name=SHEET).set_index('date')
sofrdata_raw.columns = sofrdata_raw.loc['maturity'] 
sofrdata_raw.drop(index=['maturity'],inplace=True)
sofrdata_raw.index = pd.to_datetime(sofrdata_raw.index)

sofrdata_raw.columns.name ='maturity'
sofrdata_raw /= 100
sofrdata_raw = sofrdata_raw.T.drop_duplicates().T
sofrdata = sofrdata_raw.copy()
sofrdata = compound_rate(sofrdata,BB_COMPOUND,freqcap)
sofrcurves = interp_curves(sofrdata, dt = 1/freqcap, date=DATE, interp_method='cubicspline').rename(columns={'interp':'swap rates'})
sofrquotes = sofrdata.loc[DATE,:].rename('quotes')

sofrcurves['spot rates'] = bootstrap_discounts_clean(sofrcurves[['swap rates']],compounding=4,key='swap rates')
sofrcurves['discounts'] = ratecurve_to_discountcurve(sofrcurves['spot rates'],n_compound=freqcap)
sofrcurves['forwards'] = ratecurve_to_forwardcurve(sofrcurves['spot rates'],n_compound=freqcap)

In [134]:
SHEET = 'cap'
capdata = pd.read_excel(f'{FILEIN}.xlsx', sheet_name=SHEET).set_index('date')
capdata.columns = capdata.loc['maturity'] 
capdata.drop(index=['maturity'],inplace=True)
capdata.index = pd.to_datetime(capdata.index)
capdata.columns = (freqcap * capdata.columns.values).round(0)/freqcap
capdata.columns.name ='maturity'
capdata = capdata.T.drop_duplicates().T
capquotes = capdata.loc[DATE,:].to_frame()
capquotes.columns = ['normal']
capquotes[DATE] = capquotes['normal'] / sofrcurves['forwards'] / 100**2
LIN_EXTRAP = True
FRONT_RATIO = .65

capcurves = interp_curves(capquotes[[DATE]].T,dt=1/freqcap, date=DATE,interp_method='cubicspline').rename(columns={'interp':'flat vols'})

if LIN_EXTRAP:
    fix_short = capcurves.loc[:1,'flat vols']
    fix_short.iloc[:-1] = np.nan
    fix_short.iloc[0] = capcurves.loc[1,'flat vols'] * FRONT_RATIO
    capcurves.loc[:1,'flat vols'] = fix_short.interpolate(limit_direction='both', fill_value = 'extrapolate')

# drop extrapolated value at first period as there is no caplet for the first period.
capcurves['flat vols'].iloc[0] = np.nan
capcurves = flat_to_forward_vol_rev(capcurves['flat vols'],sofrcurves['swap rates'],sofrcurves['forwards'],sofrcurves['discounts'],freq=4)
curves = pd.concat([sofrcurves.drop(columns=['quotes']), capcurves.drop(columns=['cap prices'])],axis=1)
curves.index.name = 'tenor'
curves

swap rates  spot rates  discounts  forwards  flat vols  fwd vols
tenor                                                                  
0.25     0.052211    0.052211   0.987115       NaN        NaN       NaN
0.50     0.051540    0.051535   0.974722  0.050860   0.166025  0.166025
0.75     0.050506    0.050490   0.963069  0.048400   0.191290  0.210648
1.00     0.049284    0.049250   0.952230  0.045531   0.216554  0.254312
1.25     0.047631    0.047565   0.942608  0.040831   0.260043  0.361247
1.50     0.046235    0.046141   0.933499  0.039030   0.292615  0.380930
1.75     0.045059    0.044939   0.924774  0.037738   0.315878  0.388953
2.00     0.044133    0.043994   0.916212  0.037382   0.331443  0.386643
2.25     0.043173    0.043011   0.908230  0.035151   0.340919  0.376247
2.50     0.042461    0.042283   0.900188  0.035738   0.345916  0.363764
2.75     0.041934    0.041745   0.892076  0.036373   0.348042  0.353513
3.00     0.041542    0.041347   0.883906  0.036973   0.348907  0.349207
3.25     0.041101    0.040896   0.876134  0.035482   0.349778  0.350339
3.50     0.040690    0.040473   0.868539  0.034978   0.350549  0.351556
3.75     0.040416    0.040195   0.860726  0.036309   0.350772  0.349233
4.00     0.040197    0.039974   0.852910  0.036657   0.349998  0.339501
4.25     0.039940    0.039710   0.845410  0.035486   0.347970  0.322525
4.50     0.039699    0.039461   0.838026  0.035245   0.345200  0.309744
4.75     0.039540    0.039301   0.830464  0.036421   0.342387  0.306238
5.00     0.039432    0.039195   0.822816  0.037180   0.340234  0.311414
5.25     0.039329    0.039093   0.815264  0.037055   0.339208  0.324078
5.50     0.039225    0.038990   0.807828  0.036819   0.338844  0.332488
5.75     0.039127    0.038891   0.800481  0.036713   0.338442  0.331264
6.00     0.039038    0.038802   0.793191  0.036762   0.337304  0.317965
6.25     0.038964    0.038728   0.785930  0.036954   0.334939  0.294203
6.50     0.038902    0.038668   0.778698  0.037151   0.331698  0.273998
6.75     0.038849    0.038617   0.771503  0.037305   0.328142  0.262325
7.00     0.038804    0.038574   0.764355  0.037404   0.324831  0.260895
7.25     0.038763    0.038535   0.757265  0.037451   0.322182  0.268693
7.50     0.038726    0.038501   0.750230  0.037507   0.320050  0.275091
7.75     0.038695    0.038471   0.743246  0.037589   0.318148  0.276397
8.00     0.038668    0.038447   0.736306  0.037699   0.316187  0.271592
8.25     0.038646    0.038429   0.729407  0.037833   0.313948  0.261350
8.50     0.038629    0.038415   0.722550  0.037964   0.311483  0.251699
8.75     0.038616    0.038406   0.715735  0.038085   0.308914  0.244536
9.00     0.038606    0.038400   0.708966  0.038190   0.306360  0.240251
9.25     0.038598    0.038396   0.702246  0.038278   0.303943  0.239254
9.50     0.038593    0.038395   0.695578  0.038344   0.301782  0.241978
9.75     0.038588    0.038395   0.688967  0.038385   0.300000  0.248893
10.00    0.038584    0.038395   0.682416  0.038398   0.298715  0.260524

Functions Used:

## 1.1
Calculate the (relevant) forward swap rate. That is, the one-year forward 4-year swap rate.

In [135]:
Topt = expry
Tswap = Topt + tenor

fwdrate = curves['forwards'].loc[Tswap]

fwdswap = calc_fwdswaprate(curves['discounts'], Topt, Tswap, freqswap = freqswap)

print(f"Forward swap rate: {fwdswap:.2%}")

Forward swap rate: 3.67%


In [120]:
fwdswaps

array([0.03910405, 0.03619278, 0.03743004])


## 1.2
Price the swaptions at the quoted implied volatilites and corresponding strikes, all using the just-calculated forward swap rate as the underlying.


In [136]:
voldata = pd.read_excel(f'../data/swaption_vol_data.xlsx')
volquote = voldata.query(f'model == "{QUOTE_STYLE}"').query(f'reference == "{SWAP_TYPE}"').query(f'date == "{DATE}"')
idx = (volquote['expiration'] == expry) & (volquote['tenor'] == tenor)
volquote = volquote.loc[idx]
volquote.index = ['implied vol']
strikerange = np.array(volquote.columns[-9:].tolist())
vols = volquote[strikerange]
vols /= 100
strikes = fwdswap + strikerange/100/100
idstrike = np.where(strikerange == RELATIVE_STRIKE)[0][0]
idstrikeATM = np.where(strikerange == 0)[0][0]

if QUOTE_STYLE =='normal':
    vols /= 100 * fwdrate

capvol = curves.loc[Topt, 'fwd vols']

strikeATM = strikes[idstrikeATM]
volATM = vols.iloc[0,idstrikeATM]

period_fwd = curves.index.get_loc(Topt)
period_swap = curves.index.get_loc(Tswap) + 1
step = round(freqcurve/ freqswap)

discount = curves['discounts'].iloc[period_fwd + step :period_swap:step].sum() / freqswap
black_quotes = vols.copy()
black_quotes.loc['price'] = N * blacks_formula(Topt, vols, strikes, fwdswap, discount, isCall = isPayer)[0]
black_quotes.loc['strike'] = strikes
black_quotes.style.format('{:.4f}')


## 1.3
To consider how the expiration and tenor matter, calculate the prices of a few other swaptions for comparison. 
* No need to get other implied vol quotes--just use the ATM implied vol you have for the 1x2 above. (Here we are just interested in how Black's formula changes with changes in tenor and expiration.
* No need to calculate for all the strikes--just do the ATM strike.

Alternate swaptions
* The 3mo x 4yr swaption
* The 2yr x 4yr swaption
* the 1yr x 2yr swaption

Report these values and compare them to the price of the `1y x 4y` swaption.

In [137]:
def highlight_row(row):
    if row.name == 3:
        return ['background-color: yellow'] * len(row)
    else:
        return[''] * len(row)


expiries = [.25, 2, 1, 1]
tenors = [4,4,2, 4]
black_quotes_alt = calculate_black_quotes(expiries, tenors, volATM, strikeATM, fwdswap, curves['discounts'],  isPayer)
black_quotes_alt.style.apply(highlight_row, axis =1).format({'expiry': '{:.2f}', 'tenor': '{:.2f}', 'price': '{:.4f}'})

***

# 2. Pricing w/ BDT

Use the data in `cap_curves_2024-02-20.xlsx`.

## 2.1

Calibrate the BDT Tree
* theta to fit the term structure discounts.
* sigma to fit the fwd vols from the cap data.

Report the rate tree through $T=5$. Report trees for rates compounded
* continuously
* annually

## 2.2

Use a tree to price a vanilla fixed-rate, 5-year bond with coupon rate equal to the forward swap rate calculated in problem `1.1.`

## 2.3

We will calculate the binomial tree for the 5-year swap, but here we do so by valuing the swap as...

$$\text{payer swap} = \text{floating rate note} - \text{fixed-rate bond}$$

Recall for the Floating-Rate Note:
* It has par value of 100 at each reset date.
* Every node is a reset date given the assumptions of the swap timing.

Report the tree for the 5-year swap.

## 2.4



Report the binomial tree for the one-year swaption on a 4-year swap with **european** exercise.
* At expiration, the swap tree from 2.3 will have 4 years left, as desired for pricing the 1y-4y swaption.

## 2.5

Compare the pricing of the 1y4y swaption from Black's formula in Section 1 vs the binomial tree.

## 2.6

Reprice the swaption using the BDT tree, but this time assuming it is **american**-style exercise.

***

# 3. Midcurve Swaptions

## 3.1 

Use the BDT tree from section 2 to price a **european** midcurve swaption 1y$\rightarrow$2y$\rightarrow$2y.

## 3.2

Price the **american** midcurve swaption 1y$\rightarrow$2y$\rightarrow$2y.

***